compiler/modules/front/const.bal

```
@@ -517,47 +520,94 @@ function foldVarRefExpr(FoldContext cx, t:SemType? expectedType, s:VarRefExpr ex
+    t:Context tc = cx.typeContext();
+    Position startPos = expr.startPos;
+    if expectedType == () {
+        result = check intFromLiteral(cx, expr);
+    }  
+    else if expectsInt(tc, expectedType) {
+        result = check intFromLiteral(cx, expr);
     }
-    else {
-        result = s:intFromIntLiteral(expr.base, expr.digits);
-        ty = "int";
+    else if expectsFloat(tc, expectedType) {
+        result = floatFromFpLiteral(expr.digits);
     }
-    if result is int|float {
-        return { startPos: expr.startPos, endPos: expr.endPos, value: result };
+    else if expectsDecimal(tc, expectedType) {
+        result = check decimalFromFpLiteral(cx, expr.digits, startPos);
     }
     else {
-        return cx.semanticErr("invalid " + ty + " literal", cause=result, pos=expr.startPos);
+        result = check intFromLiteral(cx, expr);
     }
+    return { startPos: startPos, endPos: expr.endPos, value: result };
 }
 
-function expectsFloat(t:SemType? semType) returns boolean {
-    if semType == () {
-        return false;
+function expectsInt(t:Context cx, t:SemType semType) returns boolean {
+    return !t:isEmpty(cx, t:intersect(semType, t:INT));
+}
+
+function expectsFloat(t:Context cx, t:SemType semType) returns boolean {
+    return !t:isEmpty(cx, t:intersect(semType, t:FLOAT));
+}
+
+function expectsDecimal(t:Context cx, t:SemType semType) returns boolean {
+    return !t:isEmpty(cx, t:intersect(semType, t:DECIMAL));
+}
+
+// Since the binary floating point literal is parsed correctly,
+// it is impossible to throw an error.
+function floatFromFpLiteral(string digits) returns float {
+    float|error f = float:fromString(digits);
+    if f is error {
+        panic err:impossible();
     }
-    else {
-        return t:isSubtypeSimple(t:intersect(semType, t:union(t:FLOAT, t:INT)), t:FLOAT);
+    return f;
+}
+
+// Even if the decimal floating point literal is parsed correctly,
+// infinities should through an error.
```

overflow should throw an error.

---

Changed